### PR TITLE
Updated Intercom Support (2024)

### PR DESCRIPTION
**Pull request by cosimomeli**
Another PR for Intercom, forked from andrew-rinato.

- Add test and test data for Intercom
- Real world test of Open Door with different user_id: I made it as a function param so a real user_id can be passed, if a fake one is provided the app just says "Shared User", without error.

Please let me know if something needs to be fixed, I promise this PR will not be abandoned. :)

### CODE REVIEW

**Comment by sdb9696** [comment]
Hi @cosimomeli, thanks for the PR and the added tests!  Could you please get the CI passing (and rebase to master) before I take a detailed look? However one quick bit of feedback is that we should probably call the class in other `RingOther` to be consistent with the other device classes.  Thanks.

File: tests/fixtures/ring_intercom_settings.json [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
...e, \"idle_lvl\": 32, \"ext_audio\": false, \"ext_audio_term\": 0, \"off_hk_tm\": 0, \"unlk_ka\": false, \"unlock\": {\"cap_tm\": 1000, \"rpl_tm\": 1500, \"gain\": 1000, \"cmp_thr\": 4000, \"lvl\": 25000, \"thr\": 30, \"thr2\": 0, \"offln\": false, \"ac\": true, \"bias\": \"h\", \"tx_pin\": \"TXD2\", \"ack\": 0, \"prot\": \"Comelit_SB2\", \"ask\": {\"f\": 25000, \"b\": 333}}, \"ring\": {\"cap_tm\": 40, \"rpl_tm\": 200, \"gain\": 2000, \"cmp_thr\": 4500, \"lvl\": 28000, \"thr\": 30, \"thr2\": 0, \"offln\": false, \"ac\": true, \"bias\": \"m\", \"tx_pin\": \"TXD2\", \"ack\": 0, \"prot\": \"Comelit_SB2\", \"ask\": {\"f\": 25000, \"b\": 333}}, \"hook_off\": {\"cap_tm\": 1000, \"rpl_tm\": 1500, \"gain\": 1000, \"cmp_thr\": 4000, \"lvl\": 25000, \"thr\": 30, \"thr2\": 0, \"offln\": false, \"ac\": true, \"bias\": \"h\", \"tx_pin\": \"TXD2\", \"ack\": 0, \"prot\": \"Comelit_SB2\", \"ask\": {\"f\": 25000, \"b\": 333}}, \"hook_on\": {\"cap_tm\": 1000, \"rpl_tm\": 1500, \"gain\": 1000, \"cmp_thr\": 4000, \"lvl\": 25000, \"thr\": 30, \"thr2\": 0, \"offln\": false, \"ac\": true, \"bias\": \"h\", \"tx_pin\": \"TXD2\", \"ack\": 0, \"prot\": \"Comelit_SB2\", \"ask\": {\"f\": 25000, \"b\": 333}}}}",
```
Comment (sdb9696) [inline tests/fixtures/ring_intercom_settings.json:362]: Is this actually how the raw data is returned from the ring api query?  With the config escaped like this?

File: ring_doorbell/other.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    def model(self):
+        """Return Ring device model name."""
+        if self.kind in INTERCOM_KINDS:
+            return "Intercom"
+        return None
+
+    def has_capability(self, capability):
+        """Return if device has specific capability."""
+        if capability == "open":
+            return self.kind in INTERCOM_KINDS
```
Comment (sdb9696) [inline ring_doorbell/other.py:61]: Is this the only capability the device has? I think it has motion detection and battery etc no?

File: ring_doorbell/other.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    @clip_length_max.setter
+    def clip_length_max(self, value):
+        url = SETTINGS_ENDPOINT.format(self.device_api_id)
+        payload = {"video_settings": {"clip_length_max": value}}
+        try:
+            self._ring.query(url, method="PATCH", json=payload)
+            self._ring.update_devices()
+            return True
+        except Exception as ex:
+            raise RingError(f"Unknown error during query of url {url}: {ex}") from ex
```
Comment (sdb9696) [inline ring_doorbell/other.py:208]: Why is this catch needed as `query` should only raise `RingError`.  Was it for debugging?

https://github.com/tchellomello/python-ring-doorbell/blob/master/ring_doorbell/auth.py#L155

**Review by sdb9696 — CHANGES_REQUESTED** [review]
Hi @cosimomeli, thanks for the great work on this PR!  A few minor point and questions in the review to be addressed/answered.

Additionally can you confirm that the list below is the complete list of fields that need to be redacted from data returned by the new intercom devices:

```
"id",
"device_id",
"description",
"first_name",
"last_name",
"email",
"location_id",
"ring_net_id",
"wifi_name",
"latitude",
"longitude",
"address",
"ring_id",
```

These are the fields currently redacted from homeassistant diagnostics dump so if there are any additional field names to be added please create a new constant called REDACT_FIELD_NAMES with the list above plus any additional and I'll switch HA to pull this from the library.

Are you a homeassistant user?  A lot of the requests for this functionality are coming from the HA community but I have no way of testing this in HA myself so it would be good if you are.  If not I can always try to find someone else to help.

File: tests/fixtures/ring_intercom_users.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
+[
+  {
+    "id": 115201490,
+    "verified": true,
+    "first_name": "John",
+    "last_name": "Carter",
+    "email": "john@cart.er",
```
[the remark this answers is not in the document]
Comment (cosimomeli) [reply tests/fixtures/ring_intercom_users.json:7]: Yes, fake name. I wrote the first thing that came to my mind, so feel free to use another fake user if needed.

File: tests/fixtures/ring_intercom_settings.json (hunk 1 above)
Comment (cosimomeli) [reply tests/fixtures/ring_intercom_settings.json:362]: Exactly, that's the raw data of my Intercom. I got it using the `raw-query` of the CLI.

File: ring_doorbell/other.py (hunk 2 above)
Comment (cosimomeli) [reply ring_doorbell/other.py:61]: The Ring Intercom does only 3 things: receives calls, makes voice communication with the connected intercom, and opens the only door connected. 
It has a battery and makes available the battery level, is this a capability to add?

File: tests/test_other.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        == "/clients_api/locations/mock-location-id/invitations/123456789"
+    )
+
+
+def test_other_open_door(ring, requests_mock):
+    dev = ring.devices()["other"][0]
+
+    import uuid
+
+    uuid.uuid4 = lambda: "987654321"
```
[the remark this answers is not in the document]
Comment (cosimomeli) [reply tests/test_other.py]: Sorry, didn't about an easy and fast way to global patch a module. Thanks!

File: ring_doorbell/other.py (hunk 2 above)
Comment (sdb9696) [reply ring_doorbell/other.py:60]: Let's leave it as is for now

**Comment by sdb9696** [comment]
Hi @cosimomeli, those tweaks look good.  Could you just come back to me the questions here https://github.com/tchellomello/python-ring-doorbell/pull/330#pullrequestreview-1845482581 and we're nearly good to go.

**Comment by cosimomeli** [comment]
> These are the fields currently redacted from homeassistant diagnostics dump so if there are any additional field names to be added please create a new constant called REDACT_FIELD_NAMES with the list above plus any additional and I'll switch HA to pull this from the library.

The class has a property `allowed_users`, with data about users with access to the device. Are `@property` included in the dump?

`REDACT_FIELD_NAMES` needs to just be defined inside `const.py` and nothing more?

**Comment by sdb9696** [comment]
> > These are the fields currently redacted from homeassistant diagnostics dump so if there are any additional field names to be added please create a new constant called REDACT_FIELD_NAMES with the list above plus any additional and I'll switch HA to pull this from the library.
> 
> The class has a property `allowed_users`, with data about users with access to the device. Are `@property` included in the dump?
> 
> `REDACT_FIELD_NAMES` needs to just be defined inside `const.py` and nothing more?

The dump is from the devices data json returned from the service calls rather than the classes themselves, so whatever the allowed_users field names are in the json.  For now all we need to do in this library is add REDACT_FIELD_NAMES to const.py with the values above and any additional ones you come up with.

**Comment by cosimomeli** [comment]
> Are you a homeassistant user? A lot of the requests for this functionality are coming from the HA community but I have no way of testing this in HA myself so it would be good if you are. If not I can always try to find someone else to help.

Yes, I am and that's also the reason I'm here. I've never built the docker image of HA from the source, but can give it a try. With the docker image available, testing the integration should not be an issue.
Worst case I should be able to mount the folder of the integration inside an already built image.

[This](https://github.com/home-assistant/core/pull/91600) is the PR that brought me here.

**Comment by cosimomeli** [comment]
> > > These are the fields currently redacted from homeassistant diagnostics dump so if there are any additional field names to be added please create a new constant called REDACT_FIELD_NAMES with the list above plus any additional and I'll switch HA to pull this from the library.
> > 
> > 
> > The class has a property `allowed_users`, with data about users with access to the device. Are `@property` included in the dump?
> > `REDACT_FIELD_NAMES` needs to just be defined inside `const.py` and nothing more?
> 
> The dump is from the devices data json returned from the service calls rather than the classes themselves, so whatever the allowed_users field names are in the json. For now all we need to do in this library is add REDACT_FIELD_NAMES to const.py with the values above and any additional ones you come up with.

If `REDACT_FIELD_NAMES` works recursively we're good to go, as the sensitive fields inside the list of `allowed_users` are already there. Same thing for the `owner` object. If it's not recursive I need to know the notation for field nested in objects and lists.

**Review by sdb9696 — APPROVED** [review]
It is recursive so I think we're all good to go here.

Are you on discord?  We may find it easier to collaborate on the HA PR that way.

**Comment by sdb9696** [comment]
> Are you on discord? We may find it easier to collaborate on the HA PR that way.

Hey @cosimomeli any thoughts here? Homeassistant uses discord for collaboration and the platform is quite effective for working on the integration libraries as well. My handle is the same there.

**Comment by cosimomeli** [comment]
> > Are you on discord? We may find it easier to collaborate on the HA PR that way.
> 
> Hey @cosimomeli any thoughts here? Homeassistant uses discord for collaboration and the platform is quite effective for working on the integration libraries as well. My handle is the same there.

Ok, we can meet on Discord. I'll ping you asap.

**Pull request merged**